### ISSUE TITLE
Bleadperl v5.13.6-543-g2d0d1ec breaks GFUJI/Class-Monadic-0.04.tar.gz

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#11227]
Migrated from [rt.perl.org#87664](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=87664) (status was 'resolved')

Searchable as RT87664$

**Comment by p5pRT** [comment]
### From @andk

rt\.cpan ticket&#8203;:


---
&nbsp;  https://rt.cpan.org/Ticket/Display.html?id=66921

git bisect&#8203;:


---
&nbsp;  2d0d1eccfcfe45f2fac66080cd30117a608088f8 is the first bad commit
&nbsp;  commit 2d0d1eccfcfe45f2fac66080cd30117a608088f8
&nbsp;  Author&#8203;: Father Chrysostomos \<sprout@&#8203;cpan\.org>
&nbsp;  Date&#8203;:   Mon Nov 15 14&#8203;:30&#8203;:07 2010 \-0800

&nbsp;  \[perl \#79208\] %stash&#8203;:: = \(\) anonymises CVs

&nbsp;  This keeps stash names visible during %foo&#8203;:: = \(\)\.

&nbsp;  This fixes @&#8203;ISA assignment inside a DESTROY method triggered by %foo&#8203;::
&nbsp;  = \(\) and also lets existing CVs retain their pointers to the stash\.

&nbsp;  So
&nbsp;    %foo&#8203;:: = \(\)
&nbsp;  is now equivalent to
&nbsp;    delete $foo&#8203;::\{$\_\} for keys %foo&#8203;::

perl \-V&#8203;:


---
&nbsp;  Summary of my perl5 \(revision 5 version 13 subversion 6\) configuration&#8203;:
&nbsp;    Commit id&#8203;: 2d0d1eccfcfe45f2fac66080cd30117a608088f8
&nbsp;    Platform&#8203;:
&nbsp;      osname=linux\, osvers=2\.6\.32\-5\-amd64\, archname=x86\_64\-linux
&nbsp;      uname='linux k81 2\.6\.32\-5\-amd64 \#1 smp wed jan 12 03&#8203;:40&#8203;:32 utc 2011 x86\_64 gnulinux '
&nbsp;      config\_args='\-Dprefix=/home/src/perl/repoperls/installed\-perls/perl/v5\.13\.6\-543\-g2d0d1ec \-Dinstallusrbinperl=n \-Uversiononly \-Dusedevel \-des \-Ui\_db \-Uuseithreads \-Uuselongdouble \-DDEBUGGING=\-g'
&nbsp;      hint=recommended\, useposix=true\, d\_sigaction=define
&nbsp;      useithreads=undef\, usemultiplicity=undef
&nbsp;      useperlio=define\, d\_sfio=undef\, uselargefiles=define\, usesocks=undef
&nbsp;      use64bitint=define\, use64bitall=define\, uselongdouble=undef
&nbsp;      usemymalloc=n\, bincompat5005=undef
&nbsp;    Compiler&#8203;:
&nbsp;      cc='cc'\, ccflags ='\-fno\-strict\-aliasing \-pipe \-fstack\-protector \-I/usr/local/include \-D\_LARGEFILE\_SOURCE \-D\_FILE\_OFFSET\_BITS=64'\,
&nbsp;      optimize='\-O2 \-g'\,
&nbsp;      cppflags='\-fno\-strict\-aliasing \-pipe \-fstack\-protector \-I/usr/local/include'
&nbsp;      ccversion=''\, gccversion='4\.5\.2'\, gccosandvers=''
&nbsp;      intsize=4\, longsize=8\, ptrsize=8\, doublesize=8\, byteorder=12345678
&nbsp;      d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=16
&nbsp;      ivtype='long'\, ivsize=8\, nvtype='double'\, nvsize=8\, Off\_t='off\_t'\, lseeksize=8
&nbsp;      alignbytes=8\, prototype=define
&nbsp;    Linker and Libraries&#8203;:
&nbsp;      ld='cc'\, ldflags =' \-fstack\-protector \-L/usr/local/lib'
&nbsp;      libpth=/usr/local/lib /lib /usr/lib /lib64 /usr/lib64
&nbsp;      libs=\-lnsl \-lgdbm \-ldb \-ldl \-lm \-lcrypt \-lutil \-lc \-lgdbm\_compat
&nbsp;      perllibs=\-lnsl \-ldl \-lm \-lcrypt \-lutil \-lc
&nbsp;      libc=/lib/libc\-2\.11\.2\.so\, so=so\, useshrplib=false\, libperl=libperl\.a
&nbsp;      gnulibc\_version='2\.11\.2'
&nbsp;    Dynamic Linking&#8203;:
&nbsp;      dlsrc=dl\_dlopen\.xs\, dlext=so\, d\_dlsymun=undef\, ccdlflags='\-Wl\,\-E'
&nbsp;      cccdlflags='\-fPIC'\, lddlflags='\-shared \-O2 \-g \-L/usr/local/lib \-fstack\-protector'


&nbsp;  Characteristics of this binary \(from libperl\)&#8203;: 
&nbsp;    Compile\-time options&#8203;: PERL\_DONT\_CREATE\_GVSV PERL\_MALLOC\_WRAP PERL\_USE\_DEVEL
&nbsp;                          USE\_64\_BIT\_ALL USE\_64\_BIT\_INT USE\_LARGE\_FILES
&nbsp;                          USE\_PERLIO USE\_PERL\_ATOF
&nbsp;    Built under linux
&nbsp;    Compiled at Apr  2 2011 18&#8203;:29&#8203;:36
&nbsp;    @&#8203;INC&#8203;:
&nbsp;      /home/src/perl/repoperls/installed\-perls/perl/v5\.13\.6\-543\-g2d0d1ec/lib/site\_perl/5\.13\.6/x86\_64\-linux
&nbsp;      /home/src/perl/repoperls/installed\-perls/perl/v5\.13\.6\-543\-g2d0d1ec/lib/site\_perl/5\.13\.6
&nbsp;      /home/src/perl/repoperls/installed\-perls/perl/v5\.13\.6\-543\-g2d0d1ec/lib/5\.13\.6/x86\_64\-linux
&nbsp;      /home/src/perl/repoperls/installed\-perls/perl/v5\.13\.6\-543\-g2d0d1ec/lib/5\.13\.6
&nbsp;      \.


\-\- 
andreas

**Comment by p5pRT** [comment]
### From @cpansprout

Can this be made into a blocker?

For that matter\, can all BBC article automatically become blockers?

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

Can this be made into a blocker?

For that matter\, can all BBC article automatically become blockers?

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @cpansprout


On Apr 2\, 2011\, at 10&#8203;:36 AM\, Andreas J\. Koenig wrote&#8203;:

> rt\.cpan ticket&#8203;:
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> 
>  https://rt.cpan.org/Ticket/Display.html?id=66921
> 
> git bisect&#8203;:
> \-\-\-\-\-\-\-\-\-\-\-
> 
>  2d0d1eccfcfe45f2fac66080cd30117a608088f8 is the first bad commit
>  commit 2d0d1eccfcfe45f2fac66080cd30117a608088f8
>  Author&#8203;: Father Chrysostomos \<sprout@&#8203;cpan\.org>
>  Date&#8203;:   Mon Nov 15 14&#8203;:30&#8203;:07 2010 \-0800
> 
>  \[perl \#79208\] %stash&#8203;:: = \(\) anonymises CVs
> 
>  This keeps stash names visible during %foo&#8203;:: = \(\)\.
> 
>  This fixes @&#8203;ISA assignment inside a DESTROY method triggered by %foo&#8203;::
>  = \(\) and also lets existing CVs retain their pointers to the stash\.
> 
>  So
>    %foo&#8203;:: = \(\)
>  is now equivalent to
>    delete $foo&#8203;::\{$\_\} for keys %foo&#8203;::
> 

It can be reduced to this&#8203;:


---
sub Foo&#8203;::bar&#8203;::DESTROY \{
&nbsp;  my $stash = delete $Foo&#8203;::\{"bar&#8203;::"\};
&nbsp;  warn join " "\, keys %Foo&#8203;::; \# Warning\, something’s wrong
&nbsp;  %$stash=\(\);
&nbsp;  warn join " "\, keys %Foo&#8203;::; \# bar&#8203;::
\}

\{ bless \[\]\, "Foo&#8203;::bar"; \}


---
What’s happening is that the %$stash assignment is causing the DESTROY glob to be freed\. So the DESTROY sub \(since it is active and therefore has an extra refcount and won’t be freed just yet\) gets ‘anonymised’\, that is\, renamed to Foo&#8203;::bar&#8203;::\_\_ANON\_\_\. \(Before 5\.13\.7 it was renamed \_\_ANON\_\_&#8203;::\_\_ANON\_\_\.\)

This is accomplished in S\_anonymise\_cv\_maybe in sv\.c\, which makes the cv \(DESTROY\) point to that glob\. So it ends up autovivifying the package\. This is what happens also in previous releases if the stash assignment is replaced with ‘delete $$stash\{$\_\} for keys %$stash’\.

I think the best way to solve this is for S\_anonymise\_cv\_maybe to use the stash’s effective name to fetch it’s \*\_\_ANON\_\_ glob\, to avoid autovivifying it\. This will guarantee that\, when it is trying to retain the pointer to the same stash \(as opposed to %\_\_ANON\_\_&#8203;::\)\, it \*will\* be using the same stash\, and not some other \(possibly autovivified\) stash that happened to have the same name\. I think this was the original intent of this code anyway\.

The user\-visible effect of this will be that\, when %$stash is no longer attached to the symbol table\,

&nbsp;  delete $$stash\{$\_\} for keys %$stash

will no longer autovivify the stash that has just been deleted\, but will rename Foo&#8203;::bar&#8203;::DESTROY to \_\_ANON\_\_&#8203;::\_\_ANON\_\_\. \(I would consider this a bug fix\, as it eliminates an existing memory leak\.\) And

&nbsp;  %$stash = \(\);

will go back to renaming DESTROY to \_\_ANON\_\_&#8203;::\_\_ANON\_\_\, as it did in 5\.13\.6 and earlier\.

Another noticeable effect is that this code&#8203;:

&nbsp;  sub Bar&#8203;::baz\{\}
&nbsp;  my $sub = \\&Bar&#8203;::baz;
&nbsp;  \*Foo&#8203;:: = delete $&#8203;::\{"Bar&#8203;::"\};
&nbsp;  delete $Foo&#8203;::\{baz\};

will cause Bar&#8203;::baz to be renamed Bar&#8203;::\_\_ANON\_\_\, as before\, but behind the scenes it will point to the stash now at %Foo&#8203;::\, instead of autoviviying a new %Bar&#8203;::\.

Attached is a patch to do this\.

**Comment by p5pRT** [comment]
### From @cpansprout

From&#8203;: Father Chrysostomos \<sprout@&#8203;cpan\.org>

Don’t autovivify stashes when anonymising CVs

This commit stops CV anonymisation from autovivifying stashes to point
to \(unless the stash is %\_\_ANON\_\_&#8203;::\)\.

If a stash has been deleted from its original position in the symbol
table\, then its HvNAME will no longer indicate where to find it\.

S\_anonymise\_cv\_maybe in sv\.c was using the HvNAME to look up \(and
autovivify\) the \*\_\_ANON\_\_ glob in the stash\, without taking into
account that it might not actually be looking in the right spot\.

So now\, after checking that the stash still has a name \(HvNAME\)\, it
uses the HvENAME to find it\. If the HvENAME is null\, which indicates
that the stash has been detached altogether\, then %\_\_ANON\_\_&#8203;:: is used\,
as happens when HvNAME is null\.

This solves a Class&#8203;::Monadic failure introduced by commit 2d0d1eccfc
\(\[perl \#79208\] %stash&#8203;:: = \(\) anonymises CVs\)\, which was included
in 5\.13\.7\.

Basically\, it can be reduced to this&#8203;:


---
sub Foo&#8203;::0xc0ffee&#8203;::DESTROY \{
&nbsp;  my $stash = delete $Foo&#8203;::\{"0xc0ffee&#8203;::"\};
&nbsp;  %$stash=\(\);
\}


---
During the DESTROY call\, the stash is deleted out from under the
DESTROY\. The stash assignment\, which no longer hides the stash’s name
temporarily while freeing entries as of 2d0d1eccfc\, ends up anonymis\-
ing the DESTROY CV\, thereby autovivifying a new Foo&#8203;::0xc0ffee&#8203;::DESTROY
glob\. So we end up with a memory leak\. 

Class&#8203;::Monadic would have run into the same memory leak in 5\.13\.6 and
earlier if it had used ‘delete $$stash\{$\_\} for keys %$stash’ to begin
with\. This commit fixes the leak for that case\, too\.

<details><summary>Inline Patch</summary>

```diff
diff --git a/sv.c b/sv.c
index 9ea0b50..447c2bc 100644
--- a/sv.c
+++ b/sv.c
@@ -5991,7 +5991,8 @@ S_anonymise_cv_maybe(pTHX_ GV *gv, CV* cv)
     }
 
     /* if not, anonymise: */
-    stash  = GvSTASH(gv) ? HvNAME(GvSTASH(gv)) : NULL;
+    stash  = GvSTASH(gv) && HvNAME(GvSTASH(gv))
+              ? HvENAME(GvSTASH(gv)) : NULL;
     gvname = Perl_newSVpvf(aTHX_ "%s::__ANON__",
 					stash ? stash : "__ANON__");
     anongv = gv_fetchsv(gvname, GV_ADDMULTI, SVt_PVCV);
diff --git a/t/op/stash.t b/t/op/stash.t
index 3fdda1f..de1297d 100644
--- a/t/op/stash.t
+++ b/t/op/stash.t
@@ -7,7 +7,7 @@ BEGIN {
 
 BEGIN { require "./test.pl"; }
 
-plan( tests => 51 );
+plan( tests => 53 );
 
 # Used to segfault (bug #15479)
 fresh_perl_like(
@@ -136,6 +136,19 @@ SKIP: {
 	is($st, q/__ANON__/, "...and an __ANON__ stash");
     }
 
+    my $sub = do {
+	package six;
+	\&{"six"}
+    };
+    my $stash_glob = delete $::{"six::"};
+    # Now free the GV while the stash still exists (though detached)
+    delete $$stash_glob{"six"};
+    $gv = B::svref_2object($sub)->GV;
+    ok($gv->isa(q/B::GV/),
+       'anonymised CV whose stash is detached still has a GV');
+    is $gv->STASH->NAME, '__ANON__',
+     'CV anonymised when its stash is detached becomes __ANON__::__ANON__';
+
     # CvSTASH should be null on a named sub if the stash has been deleted
     {
 	package FOO;
```
</details>

**Comment by p5pRT** [comment]
### From @obra




On Sun  3\.Apr'11 at 12&#8203;:44&#8203;:03 \-0700\, Father Chrysostomos via RT wrote&#8203;:
> Can this be made into a blocker?

I've made this into a blocker

> For that matter\, can all BBC article automatically become blockers?

I'm somewhat happy with the manual review and discussion process on BBCs\.
I'm also a little leery of the "100% bugward compat" precedent that
automatically setting up any BBC as a blocker would be\.

\-Jesse

**Comment by p5pRT** [comment]
### From @cpansprout

On Sun Apr 03 21&#8203;:19&#8203;:46 2011\, jesse@&#8203;bestpractical\.com wrote&#8203;:
> 
> 
> 
> On Sun  3\.Apr'11 at 12&#8203;:44&#8203;:03 \-0700\, Father Chrysostomos via RT wrote&#8203;:
> > Can this be made into a blocker?
> 
> I've made this into a blocker
> 
> > For that matter\, can all BBC article automatically become blockers?
> 
> I'm somewhat happy with the manual review and discussion process on BBCs\.
> I'm also a little leery of the "100% bugward compat" precedent that
> automatically setting up any BBC as a blocker would be\.

What about \#87702 and \#87708? The former I think is simple enough\. The
latter I would like to see fixed\, but the change required I think is a
bit intrusive at this stage\, which probably means it’s not acceptable
during code freeze\. :\-\(  What does Dave Mitchell think?

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

On Sun Apr 03 21&#8203;:19&#8203;:46 2011\, jesse@&#8203;bestpractical\.com wrote&#8203;:
> 
> 
> 
> On Sun  3\.Apr'11 at 12&#8203;:44&#8203;:03 \-0700\, Father Chrysostomos via RT wrote&#8203;:
> > Can this be made into a blocker?
> 
> I've made this into a blocker
> 
> > For that matter\, can all BBC article automatically become blockers?
> 
> I'm somewhat happy with the manual review and discussion process on BBCs\.
> I'm also a little leery of the "100% bugward compat" precedent that
> automatically setting up any BBC as a blocker would be\.

What about \#87702 and \#87708? The former I think is simple enough\. The
latter I would like to see fixed\, but the change required I think is a
bit intrusive at this stage\, which probably means it’s not acceptable
during code freeze\. :\-\(  What does Dave Mitchell think?

**Comment by p5pRT** [comment]
### From @cpansprout

On Sun Apr 03 13&#8203;:13&#8203;:09 2011\, sprout wrote&#8203;:
> 
> On Apr 2\, 2011\, at 10&#8203;:36 AM\, Andreas J\. Koenig wrote&#8203;:
> 
> > rt\.cpan ticket&#8203;:
> > \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> >
> >  https://rt.cpan.org/Ticket/Display.html?id=66921
> >
> > git bisect&#8203;:
> > \-\-\-\-\-\-\-\-\-\-\-
> >
> >  2d0d1eccfcfe45f2fac66080cd30117a608088f8 is the first bad commit
> >  commit 2d0d1eccfcfe45f2fac66080cd30117a608088f8
> >  Author&#8203;: Father Chrysostomos \<sprout@&#8203;cpan\.org>
> >  Date&#8203;:   Mon Nov 15 14&#8203;:30&#8203;:07 2010 \-0800
> >
> >  \[perl \#79208\] %stash&#8203;:: = \(\) anonymises CVs
> >
> >  This keeps stash names visible during %foo&#8203;:: = \(\)\.
> >
> >  This fixes @&#8203;ISA assignment inside a DESTROY method triggered by
>    %foo&#8203;::
> >  = \(\) and also lets existing CVs retain their pointers to the stash\.
> >
> >  So
> >    %foo&#8203;:: = \(\)
> >  is now equivalent to
> >    delete $foo&#8203;::\{$\_\} for keys %foo&#8203;::
> >
> 
> It can be reduced to this&#8203;:
> \-\-\-
> sub Foo&#8203;::bar&#8203;::DESTROY \{
>   my $stash = delete $Foo&#8203;::\{"bar&#8203;::"\};
>   warn join " "\, keys %Foo&#8203;::; \# Warning\, something’s wrong
>   %$stash=\(\);
>   warn join " "\, keys %Foo&#8203;::; \# bar&#8203;::
> \}
> 
> \{ bless \[\]\, "Foo&#8203;::bar"; \}
> \-\-\-
> 
> What’s happening is that the %$stash assignment is causing the DESTROY
>    glob to be freed\. So the DESTROY sub \(since it is active and
>    therefore has an extra refcount and won’t be freed just yet\) gets
>    ‘anonymised’\, that is\, renamed to Foo&#8203;::bar&#8203;::\_\_ANON\_\_\. \(Before
>    5\.13\.7 it was renamed \_\_ANON\_\_&#8203;::\_\_ANON\_\_\.\)
> 
> This is accomplished in S\_anonymise\_cv\_maybe in sv\.c\, which makes the
>    cv \(DESTROY\) point to that glob\. So it ends up autovivifying the
>    package\. This is what happens also in previous releases if the
>    stash assignment is replaced with ‘delete $$stash\{$\_\} for keys
>    %$stash’\.
> 
> I think the best way to solve this is for S\_anonymise\_cv\_maybe to use
>    the stash’s effective name to fetch it’s \*\_\_ANON\_\_ glob\, to avoid
>    autovivifying it\. This will guarantee that\, when it is trying to
>    retain the pointer to the same stash \(as opposed to %\_\_ANON\_\_&#8203;::\)\,
>    it \*will\* be using the same stash\, and not some other \(possibly
>    autovivified\) stash that happened to have the same name\. I think
>    this was the original intent of this code anyway\.
> 
> The user\-visible effect of this will be that\, when %$stash is no
>    longer attached to the symbol table\,
> 
>   delete $$stash\{$\_\} for keys %$stash
> 
> will no longer autovivify the stash that has just been deleted\, but
>    will rename Foo&#8203;::bar&#8203;::DESTROY to \_\_ANON\_\_&#8203;::\_\_ANON\_\_\. \(I would
>    consider this a bug fix\, as it eliminates an existing memory leak\.\)
>    And
> 
>   %$stash = \(\);
> 
> will go back to renaming DESTROY to \_\_ANON\_\_&#8203;::\_\_ANON\_\_\, as it did in
>    5\.13\.6 and earlier\.
> 
> Another noticeable effect is that this code&#8203;:
> 
>   sub Bar&#8203;::baz\{\}
>   my $sub = \\&Bar&#8203;::baz;
>   \*Foo&#8203;:: = delete $&#8203;::\{"Bar&#8203;::"\};
>   delete $Foo&#8203;::\{baz\};
> 
> will cause Bar&#8203;::baz to be renamed Bar&#8203;::\_\_ANON\_\_\, as before\, but behind
>    the scenes it will point to the stash now at %Foo&#8203;::\, instead of
>    autoviviying a new %Bar&#8203;::\.
> 
> Attached is a patch to do this\.
> 

Now applied as 57f45d7b\.

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'resolved'
